info: support stack unwinding

shintaro-iwasaki opened the pull request:
This PR implements a stack unwinding feature in Argobots. This PR should address #101 pointed out by @bfaccini .

## Why it is useful

This stack unwinding feature is for debugging. Specifically, this feature enables a user to dump thread information including ULTs in pools in a more user-friendly manner.

## How to use it

1. Configure Argobots with libunwind (I checked this with libunwind-1.4.0, but it should work with LLVM's libunwind).
```sh
./configure --with-libunwind=LIBUNWIND_INSTALL_PATH --enable-stack-unwind
```
2. Run `ABT_info_print_thread_stack_unwind()` or `ABT_info_trigger_print_all_thread_stacks()`
See https://github.com/shintaro-iwasaki/argobots/blob/supportlibunwind/test/basic/info_stack_unwind.c .

It will print the following (`ABT_info_trigger_print_all_thread_stacks()`), where `thread_func` is a user function. It will show the full stack trace if user functions are nested.
```
=== ULT (0x7f9d8888c780) ===
id        : 10
ctx       : 0x7f9d8888c7e0
  p_ctx    : 0x7f9d8888c4d0
  p_link   : (nil)
stack     : 0x7f9d88888780
stacksize : 16384
#0 0x7f9d8a2b1fc0 in ythread_unwind_stack () <+16> (RSP = 0x7f9d8888c500)
#1 0x7f9d8a2af0ed in ABT_thread_yield () <+189> (RSP = 0x7f9d8888c510)
#2 0x557292d28f8d in thread_func () <+189> (RSP = 0x7f9d8888c530)
#3 0x7f9d8a2b3ece in ABTD_ythread_func_wrapper () <+30> (RSP = 0x7f9d8888c770)
#4 0x7f9d8a2b4071 in make_fcontext () <+33> (RSP = 0x7f9d8888c780)
```

## Performance Impact
By default, this feature is disabled, so this PR does not degrade the performance.

**Enabling this feature degrades the performance on _a non-x86/64 machine_** because I implemented an optimized `peek_fcontext` assembly function only for x86/64. Precisely speaking, the performance is degraded if the underlying architecture is not (x86/64 + fcontext). If ucontext is chosen on x86/64 (= when you manually set `--disable-fcontext`), however, the performance is originally bad, so the overhead added by this feature is almost negligible.

### Q & A
Q: Can't it show code lines and/or values of arguments like GDB when a program is compiled with `-g`?
A: It is not possible with `libunwind`: it needs `dwarfdump` etc. This future work is of low priority. Please use GDB if needed (I guess, with GDB, one can check the stack trace by manually setting RSP and RIP).

Q: Why is a special optimization available only for x86/64?
A: It is because I did not implement `peek_fcontext` for other architectures. There is no fundamental technical difficulty (debugging is tedious, though), so I might support them in the future. Note that, apart from performance, this PR covers all the combinations of architectures and context types (ucontext/fcontext).

shintaro-iwasaki commented:
@bfaccini This is a prototype. I'd be happy if you could take a look at this PR and let me know if I can improve the content of printed information.

shintaro-iwasaki commented:
@bfaccini This is just a friendly reminder, but this PR implements a prototype for stack unwinding. Does this output look okay for you? Do you need more flexible/low-level interface for stack unwinding (e.g., printing register values of ULTs instead of stack trace)? I would be very happy if you could let me know any feedback.

bfaccini commented:
Hello,
Sorry to be late on this.
I am presently testing the feature, will let you know asap on how it works.
Concerning also printing registers, I would say yes it will also be helpful to get them in addition, in fact anything to facilitate debugging is important for us!
Cheers,
Bruno.

From: Shintaro Iwasaki <notifications@github.com>
Reply to: pmodels/argobots <reply@reply.github.com>
Date: Sunday 27 September 2020 at 22:31
To: pmodels/argobots <argobots@noreply.github.com>
Cc: Bruno Faccini <bruno.faccini@intel.com>, Mention <mention@noreply.github.com>
Subject: Re: [pmodels/argobots] info: support stack unwinding (#246)


@bfaccini<https://github.com/bfaccini> This is just a friendly reminder, but this PR implements a prototype for stack unwinding. Does this output look okay for you? Do you need more flexible/low-level interface for stack unwinding (e.g., printing register values of ULTs instead of stack trace)? I would be very happy if you could let me know any feedback.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pmodels/argobots/pull/246#issuecomment-699684675>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADEWAWTRUZ7P6SYCBQLMWLTSH6OK5ANCNFSM4RDXXA2A>.
---------------------------------------------------------------------
Intel Corporation SAS (French simplified joint stock company)
Registered headquarters: "Les Montalets"- 2, rue de Paris, 
92196 Meudon Cedex, France
Registration Number:  302 456 199 R.C.S. NANTERRE
Capital: 4,572,000 Euros

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

shintaro-iwasaki commented:
@bfaccini Thank you for taking a look at this PR.  I will update this PR based on your comments in a week.

> Seems that as per my tests, there are cases where the last ULT stack to be dumped is not fully available in log, so may also be the time to flush ?

I am interested in your case.  I would be happy if you could let me know what is "the last ULT stack" and how the information is "not fully available"?

Note that the current mechanism (`ABT_info_print_thread_stack()` and `ABT_info_trigger_print_all_thread_stacks()`) prints all "ready" ULT stacks in pools, so it does not print the following type of ULTs: 

- Presently running ULTs. It is difficult/complicated to call libunwind functions on top of presently running ULTs or print a raw stack of running ULTs because of thread/signal safety.

- The main ULT (=a thread that calls `ABT_init()`).  The main ULT uses Pthreads' stack, so Argobots does not know the boundary of its stack.  Thus, currently Argobots does not print the main ULT's stack.

- ULTs that are not in pools.  ULTs that are waiting on synchronization objects (e.g., mutex, barrier, ...) are detached from Argobots pools.  Those ULTs are difficult to be captured by `ABT_info_trigger_print_all_thread_stacks()`.

I believe all of the issues above can be addressed at the cost of performance (i.e., thread creation and scheduling performance).

shintaro-iwasaki commented:
Thank you! I got it. I will take a look at it and fix it in the updated PR.

> Well I am pretty sure this kind of presently missing ULTs infos will be needed at some point of time to troubleshoot very tricky scenarios.

If they become necessary, we are happy to implement them in the future although this PR will solely focus on stack unwinding. In any case, only the concern is that those enhancements would degrade the performance of Argobots, so I hesitate to enable it by default. (This unwinding PR is okay since it does not lower the basic performance of Argobots.)

shintaro-iwasaki commented on configure.ac:
```diff
@@ -722,6 +732,18 @@ if test "x$with_hugetlbfs" != "x"; then
+    # --with-libunwind
+    if test "x$with_libunwind" != "x"; then
+        PAC_PREPEND_FLAG([-I${with_libunwind}/include], [CFLAGS])
+        PAC_PREPEND_FLAG([-Wl,-rpath,${with_libunwind}/lib -L${with_libunwind}/lib -lunwind], [LDFLAGS])
```
You're right. I fixed this issue. I confirmed that Argobots works with both system libunwind (libunwind-dev) and locally installed libunwind.

shintaro-iwasaki commented on src/info.c:
```diff
@@ -803,7 +849,11 @@ static void info_print_unit(void *arg, ABT_unit unit)
+        ABTI_ythread_print_stack_unwind(p_ythread, fp);
+#else
         ABTI_ythread_print_stack(p_ythread, fp);
+#endif
```
Now these stack functions print both (dumping the full stack and libunwind information) if this feature is enabled.
I could not reproduce the latter issue. I will discuss it in the main thread.

shintaro-iwasaki replied:
Examples:
```sh
$ # ABT_info_trigger_print_all_thread_stacks
$ ./info_stackdump
...
=== ULT (0x55f0d40b4d00) ===
id        : 13
ctx       : 0x55f0d40b4d60
p_ctx    : 0x55f0d40b4a20
p_link   : (nil)
stack     : 0x55f0d40accd0
stacksize : 32768
#0 0x7f6b11490e00 in ythread_unwind_stack () <+16> (RSP = 0x55f0d40b4a50)
#1 0x7f6b1148dfbd in ABT_thread_yield () <+189> (RSP = 0x55f0d40b4a60)
#2 0x55f0d310af8d in thread_func () <+189> (RSP = 0x55f0d40b4a80)
#3 0x7f6b11492d2e in ABTD_ythread_func_wrapper () <+30> (RSP = 0x55f0d40b4cc0)
#4 0x7f6b11492ed1 in make_fcontext () <+33> (RSP = 0x55f0d40b4cd0)
000055f0d40accd0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
000055f0d40accf0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
000055f0d40acd10: 0000000000000000 00000000000202f1 000055f0d40acbd0 00007f6b11440be0
000055f0d40acd30: 0000000000000020 0000000000000050 000055f0d40acd80 0000000000008000
...
```

```sh
$ # ABT_info_print_thread_stack
$ ./info_print_stack
...
threads[3]:
p_ctx    : 0x7fc5d53985e0
p_link   : (nil)
stack     : 0x7fc5d53946c0
stacksize : 16384
#0 0x7fc5d6ff3e00 in ythread_unwind_stack () <+16> (RSP = 0x7fc5d5398610)
#1 0x7fc5d6ff0fbd in ABT_thread_yield () <+189> (RSP = 0x7fc5d5398620)
#2 0x55d29ca80dd5 in user_thread_func_lv4 () <+37> (RSP = 0x7fc5d5398640)
#3 0x55d29ca80e77 in user_thread_func_lv3 () <+103> (RSP = 0x7fc5d5398650)
#4 0x55d29ca80ef7 in user_thread_func_lv2 () <+103> (RSP = 0x7fc5d5398670)
#5 0x55d29ca80f77 in user_thread_func () <+103> (RSP = 0x7fc5d5398690)
#6 0x7fc5d6ff5d2e in ABTD_ythread_func_wrapper () <+30> (RSP = 0x7fc5d53986b0)
#7 0x7fc5d6ff5ed1 in make_fcontext () <+33> (RSP = 0x7fc5d53986c0)
00007fc5d53946c0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
00007fc5d53946e0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
00007fc5d5394700: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
...
```

shintaro-iwasaki commented on src/info.c:
```diff
@@ -494,6 +505,41 @@ int ABT_info_print_thread_stack(FILE *fp, ABT_thread thread)
+ * @return Error code
+ * @retval ABT_SUCCESS on success
+ */
+int ABT_info_print_thread_stack_unwind(FILE *fp, ABT_thread thread)
```
`ABT_info_print_thread_stack()` prints both libunwind information and a raw stack. `ABT_info_print_thread_stack_unwind()` was removed.
```sh
$ # ABT_info_print_thread_stack
$ ./info_print_stack
...
threads[3]:
p_ctx    : 0x7fc5d53985e0
p_link   : (nil)
stack     : 0x7fc5d53946c0
stacksize : 16384
#0 0x7fc5d6ff3e00 in ythread_unwind_stack () <+16> (RSP = 0x7fc5d5398610)
#1 0x7fc5d6ff0fbd in ABT_thread_yield () <+189> (RSP = 0x7fc5d5398620)
#2 0x55d29ca80dd5 in user_thread_func_lv4 () <+37> (RSP = 0x7fc5d5398640)
#3 0x55d29ca80e77 in user_thread_func_lv3 () <+103> (RSP = 0x7fc5d5398650)
#4 0x55d29ca80ef7 in user_thread_func_lv2 () <+103> (RSP = 0x7fc5d5398670)
#5 0x55d29ca80f77 in user_thread_func () <+103> (RSP = 0x7fc5d5398690)
#6 0x7fc5d6ff5d2e in ABTD_ythread_func_wrapper () <+30> (RSP = 0x7fc5d53986b0)
#7 0x7fc5d6ff5ed1 in make_fcontext () <+33> (RSP = 0x7fc5d53986c0)
00007fc5d53946c0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
00007fc5d53946e0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
00007fc5d5394700: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
...

shintaro-iwasaki commented:
@bfaccini Thank you! I believe the latest commits address the issues you pointed out (except for "the last ULT"). I'd be extremely happy if you could check these updates.

I could not reproduce the last ULT issue you found. Specifically, I ran `test/basic/info_stackdump`, `test/basic/info_stackdump2`, and `test/basic/info_print_stack` several times, but the last ULT information did not look truncated. If the issue is not caused by limitations I mentioned above, perhaps it is caused by lack of `fflush()`. However, I would like to add it after I confirm when it happens and `fflush()` (or another fix) really solves the issue. Please let me know if you know how to reproduce this issue.

bfaccini commented:
> @bfaccini Thank you! I believe the latest commits address the issues you pointed out (except for "the last ULT"). I'd be extremely happy if you could check these updates.

Seems that this new version has introduced some new regression as any attempt to use/call ABT_info_print_thread_stack(), for the 1st xstream's progress ULT, in "unattended" mode triggers a SEGV with multiple/different signatures . This was not failing with the previous version of your patch (unfortunately I can't access the patch history as you seem to have forced-push the new version). The way I implemented this "unattended" mode stacks dump in our server's main thread is to monitor/sigwait() SIGUSR1 and then enter a loop to dump stacks of all ULTs we can get reference to, either using ABT_info_print_thread_stack() for special-purpose tasks or ABT_sched_get_pools()/ABT_info_print_thread_stacks_in_pool() for all tasks in pools.
I am presently trying to gather debug infos from corefiles and will provide details asap.

> I could not reproduce the last ULT issue you found. Specifically, I ran `test/basic/info_stackdump`, `test/basic/info_stackdump2`, and `test/basic/info_print_stack` several times, but the last ULT information did not look truncated. If the issue is not caused by limitations I mentioned above, perhaps it is caused by lack of `fflush()`. However, I would like to add it after I confirm when it happens and `fflush()` (or another fix) really solves the issue. Please let me know if you know how to reproduce this issue.

I think I have found why the expected output after calling ABT_info_trigger_print_all_thread_stacks() can appear truncated during my tests. It should come from the fact output stream is not being closed as in our case the calling thread (our main server thread monitoring USR2 signals to also launch stacks dump, ie not only upon fatal signals...) don't know when to close it because the stacks dump is running asynchronously and then our server application continues running. 
I can now confirm that the missing stacks datas were not truncated but only buffered, as they become available if a second SIGUSR2 signal is being sent !!
Thus, I think that a fflush(print_stack_fp) call at the end of ABTI_info_check_print_all_thread_stacks() is required if we want all stacks dump datas to be available in output file in all cases.

shintaro-iwasaki commented:
Thank you for your feedback.

> The way I implemented this "unattended" mode stacks dump in our server's main thread is to monitor/sigwait() SIGUSR1 and then enter a loop to dump stacks of all ULTs we can get reference to, either using ABT_info_print_thread_stack() for special-purpose tasks or ABT_sched_get_pools()/ABT_info_print_thread_stacks_in_pool() for all tasks in pools.

I guess `ABT_info_print_thread_stack()` is used for tasks that are currently running, maybe caused at https://github.com/pmodels/argobots/pull/246/files#diff-8bfe01cf959cd851f68282313fb6774128e6498bb6312e73fc9a87075a4c2b77R109-R113.  Unwinding function traces of running ULTs are currently prohibited (and there is no easy way to allow this).  I added some basic checks, but this does not avoid 100% of similar failures.

>  I think that a fflush(print_stack_fp) call at the end of ABTI_info_check_print_all_thread_stacks() is required if we want all stacks dump datas to be available in output file in all cases.

Thanks! I missed `fflush()`. I added it.

It would be complicated, but what I did is 1. rebasing the existing three commits onto the `main` branch and 2. adding two commits (https://github.com/pmodels/argobots/pull/246/commits/27187b78663fbe130d30e59e5245e7225e59c87a and https://github.com/pmodels/argobots/pull/246/commits/29e6d360076e95987db3cf01409026469bb670cf) to address these issues.

( To trigger CI tests: test:argobots/all )

shintaro-iwasaki commented:
test:argobots/all

bfaccini commented:
Hello Shintaro,
I can confirm that the crashes I got and reported before when using/calling ABT_info_print_thread_stack() unattended have occurred because of the unwind attempt concurrently with the fact the same ULT was scheduled to run.
I have done more tests with the last code for this PR, but seems that even with your recent change to check if the ULT is idle before attempting to unwind its stack ([27187b7](https://github.com/pmodels/argobots/commit/27187b78663fbe130d30e59e5245e7225e59c87a)), I still get crashes, which I can say after their core files analysis, are due to the same reason.
So, I will stop using this method until we find a more solid way to fix it (may be we could simply wait it to become idle and then suspend the ULT during its stack unwinding?) and only rely on ABT_info_trigger_print_all_thread_stacks() attended way which works fine!
BTW, I can also confirm that your 2nd change ([29e6d36](https://github.com/pmodels/argobots/commit/29e6d360076e95987db3cf01409026469bb670cf)) has fixed the the output pushing issue!
Do you have an idea about when all the changes for this PR will land now ?? Also will the unwinding feature still require to build Argobots from source ?
But anyway, thanks for your help and awesome work on this !!

shintaro-iwasaki commented:
@bfaccini Thank you for checking. If this PR is somehow useful on your side, I will merge this PR.  I need to rebase this PR and set up CI for libunwind, so please give me a few days.

> Unwinding stacks safely with `ABT_info_print_thread_stack()`

Yes, unwinding stacks of running ULTs is erroneous.  Technically, we can suspend running ULTs by a signal (as done by debuggers), but invoking signals has many side effects (e.g., disrupting system calls and deadlock issues) of which users need to be fully aware.

> Will the unwinding feature still require to build Argobots from source ?

Yes.  You need to build Argobots from source to set up this feature.

shintaro-iwasaki commented:
test:argobots/all

shintaro-iwasaki commented:
I will merge this PR.  @bfaccini Thank you very much!
